python/ray/train/v2/trainer.py

```diff
@@ -0,0 +1,327 @@
+import abc
+import dataclasses
+import logging
+import os
+from typing import Dict, Union, Callable, Optional, Type
+
+import ray
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.config import ScalingConfig, RunConfig, DataParallelScalingConfig
+from ray.ml.result import Result
+from ray.train.backend import BackendExecutor
+from ray.train.checkpoint import TuneCheckpointManagerV2
+from ray.train.trainer import BACKEND_ENV_VARS
+from ray.train.utils import RayDataset
+from ray.train import BackendConfig, TrainingIterator
+from ray.train.constants import TRAIN_DATASET_KEY
+from ray.train.utils import construct_train_func
+from ray import tune
+from ray.tune import PlacementGroupFactory
+from ray.tune.function_runner import wrap_function
+from ray.tune.trainable import ConvertibleToTrainable, Trainable
+from ray.tune.utils import deep_update
+from ray.util.annotations import DeveloperAPI
+
+GenDataset = Union[RayDataset, Callable[[], RayDataset]]
+
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: support conditional hyperparameter tuning of Trainer __init__ args.
+class Trainer(ConvertibleToTrainable, abc.ABC):
+    """
+    Defines interface for distributed training on Ray.
+
+    Args:
+        scaling_config (Optional[ScalingConfig]): Configuration for how to
+            scale training.
+        run_config (Optional[RunConfig]): Configuration for the execution of
+            the training run.
+        train_dataset (Optional[GenDataset]): Either a distributed Ray
+            :ref:`Dataset <dataset-api>` or :ref:`DatasetPipeline
+            <dataset-pipeline-api>`, or a Callbable that returns a Dataset, to
+            use for training. If a ``preprocessor`` is also provided,
+            it will be fit on this dataset.
+        additional_datasets (Optional[GenDataset]): Any additional Ray
+            Datasets (such as validation or test datasets) to use for
+            training. If a ``preprocessor`` is provided, it will only
+            transform these datasets.
```

Comment:
So it will not transform the training dataset anymore? Why?

Reply:
it should fit_transform it

Reply:
Ah actually now I understand this comment. I think the implementation is correct (only call `fit_transform` on the training dataset), but the comment is confusing. Maybe explicitly say it does not fit the processor to these datasets.